### DISCUSSION TITLE
trivy not detecting dependencies defined in .csProject files (.Net)

### DISCUSSION THREAD

**Discussion by nartreb** [aquasecurity/trivy#7026] [Q&A]
### Description

A client of mine ran a trivy scan on a .Net project and didn't see output (vulns, licenses) for dependencies I know the project is using.  

If I read the documentation correctly, Trivy looks at packages.config or (preferably) packages.lock.json to tell it what dependencies are included in a .Net project.   But this project uses PackageReference tags inside a .csproject file instead (example in attached screenshot).  

![packageReference_example](https://github.com/aquasecurity/trivy/assets/35303958/fb78eb25-b6f2-4655-af4e-f75a7273a557)

### Desired Behavior

Trivy should detect all listed project dependencies, even if they are listed as PackageReferences in the .csproject file and not packages.config

(I am told that PackageReference will be the default for .Net projects:  
"This project targets .NET8 for Android which is the current/future LTS version for Microsoft (with .NET6 going end-of-life for support in November).

By default, .NET Core uses the PackageReference system, so going forward I would expect most Visual Studio projects would look this way." )

### Actual Behavior

No output reflecting those dependencies - I'm pretty sure Trivy never looked at them (I suppose maybe they're all vuln-free and none has any license metadata, but that seems unlikely.  It's not just one project or the few files in the screenshot.)

### Reproduction Steps

```bash
1. Create a .Net project.
2. Download some dependencies, (store them in a folder that is not a subdirectory of the project but is on the system path for DLLs) list them in .csproject 
  (Pick dependencies known to be detected - license metadata and/or known vulns)
3. run Trivy -fs on the project (license or vuln, as appropriate)
4.  Observe that Trivy does not report any issues related to those dependencies
5.  Also note Number of language-specific-files = 0 in the debug log... I think that means it looked at zero files
```

### Target

Filesystem

### Scanner

Vulnerability

### Output Format

Table

### Mode

Standalone

### Debug Output

```bash
2024-06-26T08:43:52-05:00	INFO	Vulnerability scanning is enabled
2024-06-26T08:43:52-05:00	INFO	Number of language-specific files	num=0
```

### Operating System

Windows

### Version

```bash
0.51.2 or thereabouts
```

### Checklist

- [X] Run `trivy clean --all`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by DmitriyLewen** [comment]
Trivy doesn't scan `.csproj` files (https://aquasecurity.github.io/trivy/v0.52/docs/coverage/language/dotnet/).

There is problem with scanning `.csproj` files.
These files can use version range - https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-version

We can't correcntly detect versions for  dependencies with version range.
See https://github.com/aquasecurity/go-dep-parser/pull/121#issuecomment-1234404017

Trivy supports the following files - https://aquasecurity.github.io/trivy/v0.52/docs/coverage/language/dotnet/
You may be able to choose a different file for your project.

Regards, Dmitriy

**Comment by nartreb** [comment]
Thanks for response.   My use case involves scanning projects that I didn't create and know absolutely nothing about.   "Choose a different file" as a prerequisite for running a scan seems backwards to me.  I need a tool that will tell me what's in the build.  A tool that only works after I've changed the build is not very useful.

I realize that not knowing exact versions of dependencies is fairly useless for security scanning, but it's much less important for license scanning, which is my top concern.

If the scanner  finds a BOM file that it can't read, I want it to warn me, so I can investigate.  And if it can sort-of-read it, I want it to warn me that results are not complete, AND show me the information it did find.   In particular, the license scan should still work even if the security scan isn't much use.

I need to be able to tell clients "just run this tool and send me the output."  If there's ANY technical steps involved beyond that, it makes my life vastly more difficult.   I *might* be able to ask my clients to  create packages.lock.json (that sounds harmless to their build) but it's decidedly sub-optimal.  

I realize you can't be expected support every weird build system that's out there.  But this is a very normal method of building .Net projects, and shouldn't be passed over in silence.

Pull #121 is a couple of years old now, I don't know if it's still valid, but it looks like it was stopped only for fear of giving users false confidence.  @knqyf263  any chance of revisiting that decision?    To me it's worse, in terms of false confidence, to completely ignore a valid BOM file.  The scan result gave me no warning about that, I only got suspicious when the number of packages seemed low.

**Reply by knqyf263** [reply]
>I realize that not knowing exact versions of dependencies is fairly useless for security scanning, but it's much less important for license scanning, which is my top concern.

It sounds reasonable to add license support, but I didn't find licenses in `.csproj`. Where can we get the license information?

**Reply by nartreb** [reply]
Just listing out the dependencies with "license: UNKNOWN" would be a huge help to me.  I can google their licenses myself.
(In fact I would run a post-processing script with a lookup table, so I would google less and less each time I scan a project.)


I'm not a .Net developer myself, just have a client who wants to audit a .Net codebase.  All I know about .csproject files is what I read here:  https://learn.microsoft.com/en-us/nuget/concepts/package-versioning?tabs=semver20sort   There's a passing mention in there of .nuspec and packages.config files, which seem like possibly useful places to look for exact version numbers (but might not exist in every project?).  Oh, the next page mentions projects.assets.json, that might be even more interesting (but again, it's not entirely clear whether that file will always exist).  But to answer your question to the best of my knowledge, none of these configuration files seems to contain license data.

It looks like PackageReference tags (inside " .csproj or other package version location (Directory.Packages.props) ") is now the preferred configuratoin method (replacing packages.config) but not fully supported yet. (Gotta love Microsoft.)  
_By default, PackageReference is used for .NET Core projects, .NET Standard projects, and UWP projects targeting Windows 10 Build 15063 (Creators Update) and later, with the exception of C++ UWP projects. .NET Framework projects support PackageReference, but currently default to packages.config. ... ASP.NET apps targeting the full .NET Framework include only [limited support](https://github.com/NuGet/Home/issues/5877) for PackageReference. C++ and JavaScript project types are unsupported._

**Reply by knqyf263** [reply]
Showing unknown versions and unknown licenses for all packages just confuses most users even if it's enough for you. It's better to say we don't support `.csproj`.

**Reply by nartreb** [reply]
Fair enough, I realize my use case is a little unusual.  But shouldn't trivy detect the presence of .csproj and put some kind of warning in the output (or standard error) ?

**Reply by knqyf263** [reply]
Since there are many unsupported ecosystems and files, it is not practical to issue warnings for all of them. It would be confusing to give special treatment to .csproj.

One thing we can do is get libraries with pinned versions only and mark others as unknown versions as we do for [requirements.txt](https://aquasecurity.github.io/trivy/v0.53/docs/coverage/language/python/#dependency-detection).

@DmitriyLewen Do you know if .csproj can pin the version?

**Reply by DmitriyLewen** [reply]
> Do you know if .csproj can pin the version?

Do you mean is there equivalent to `pip freeze` command?

**Reply by knqyf263** [reply]
Ah, we should need to know two things.

1. whether .csproj can pin the version, like `django==3.2.3`
2. whether .crproj can pin all versions, like `pip freeze`

**Reply by DmitriyLewen** [reply]
> whether .csproj can pin the version, like django==3.2.3

`.csproj` supports `==` (or `[x.y.z]`, see https://stackoverflow.com/a/78085877). But i am not sure how often it is used. (see https://learn.microsoft.com/en-us/nuget/concepts/package-versioning?tabs=semver20sort#version-ranges)
It is also necessary to understand that  `=` is not equal `==` (`=` is `>=`) (see https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-version)

> whether .crproj can pin all versions, like pip freeze

I haven't seen such functionality.
And if you consider that you can create lock file, I think there is no such command.

**Reply by knqyf263** [reply]
>But i am not sure how often it is used. 

I'm worried that all versions will be unknown.

**Reply by DmitriyLewen** [reply]
> I'm worried that all versions will be unknown.

I guess that's how it will be.
I haven't worked much with C#, but I haven't seen "==" in `.csproj` files.

**Reply by knqyf263** [reply]
Hmm. It doesn't make sense to say Trivy supports .csproj, then.
